src/datasets/iterable_dataset.py

```diff
@@ -0,0 +1,439 @@
+            function (:obj:`Callable`, optional, default None): if not None, this function is applied
+                on-the-fly on the examples when you iterate on the dataset.
+            batched (:obj:`bool`, default `False`): Provide batch of examples to `function`.
+            batch_size (:obj:`int`, optional, default ``1000``): Number of examples per batch provided to `function` if `batched=True`.
+
+        """
+        info = copy.deepcopy(self._info)
+        info.features = None
+        ex_iterable = MappedExamplesIterable(
+            self._ex_iterable, function=function, batched=batched, batch_size=batch_size
+        )
+        return iterable_dataset(
+            ex_iterable=ex_iterable,
+            info=info,
+            split=self._split,
+            format_type=self._format_type,
+            shuffling=copy.deepcopy(self._shuffling),
+        )
+
+    def shuffle(self, buffer_size, seed=None) -> "IterableDataset":
+        """
+        Randomly shuffles the elements of this dataset.
+
+        This dataset fills a buffer with buffer_size elements, then randomly samples elements from this buffer,
+        replacing the selected elements with new elements. For perfect shuffling, a buffer size greater than or
+        equal to the full size of the dataset is required.
+
+        For instance, if your dataset contains 10,000 elements but buffer_size is set to 1,000, then shuffle will
+        initially select a random element from only the first 1,000 elements in the buffer. Once an element is
+        selected, its space in the buffer is replaced by the next (i.e. 1,001-st) element,
+        maintaining the 1,000 element buffer.
+
+        Args:
+
+            buffer_size (:obj:`int`): size of the buffer.
+            seed (:obj:`int`, optional, default None): random seed that will be used to create the distribution.
+        """
+        shuffling = ShuffingConfig(seed=seed)
+        return iterable_dataset(
+            ex_iterable=BufferShuffledExamplesIterable(self._ex_iterable, buffer_size, seed=seed),
+            info=copy.deepcopy(self._info),
+            split=self._split,
+            format_type=self._format_type,
+            shuffling=shuffling,
+        )
+
+    def set_epoch(self, epoch: int):
+        self._epoch = epoch
+
+
+def iterable_dataset(
+    ex_iterable: Iterable,
+    info: Optional[DatasetInfo] = None,
+    split: Optional[NamedSplit] = None,
+    format_type: Optional[str] = None,
+    shuffling: Optional[ShuffingConfig] = None,
+):
+    if format_type is not None and format_type == "torch":
+        import torch
+
+        class TorchIterableDataset(IterableDataset, torch.utils.data.IterableDataset):
+            pass
+
+        cls = TorchIterableDataset
+    else:
+        cls = IterableDataset
+    return cls(
+        ex_iterable=ex_iterable,
+        info=info,
+        split=split,
+        format_type=format_type,
+        shuffling=shuffling,
+    )
+
+
+def interleave_datasets(
```

Comment:
I was thinking that the idea is to eventually have this function supporting regular Datasets. So maybe it should be moved to its own module?

Reply:
I moved it to a `combine` module. Maybe `concatenate_datasets` can be moved there as well